e2e/libs/volume/crd.py

```diff
@@ -30,91 +30,258 @@ def get(self, volume_name):
+                            "attachmentTickets": {
+                                "": {
+                                    "generation": 0,
+                                    "id": "",
+                                    "nodeID": node_name,
+                                    "parameters": {
+                                        "disableFrontend": "false",
+                                        "lastAttachedBy": ""
+                                    },
+                                    "type": "longhorn-api"
+                                }
+                            }
+                        }
+                }
+            )
+        except ApiException as e:
+            # new CRD: volumeattachments was added since from 1.5.0
+            # https://github.com/longhorn/longhorn/issues/3715
+            if e.reason != "Not Found":
+                Exception(f'exception for creating volumeattachments:', e)
+
         self.wait_for_volume_state(volume_name, "attached")
 
     def wait_for_volume_state(self, volume_name, desired_state):
-        print(self.get(volume_name)["status"]["state"])
+        logging.info(
+            f"waiting volume {volume_name} state becomes {desired_state}")
         for i in range(retry_count):
-            if self.get(volume_name)["status"]["state"] == desired_state:
+            volume_state = self.get(volume_name)["status"]["state"]
+            logging.debug(f"volume state: {volume_state}")
+            if volume_state == desired_state:
                 break
             time.sleep(retry_interval)
         assert self.get(volume_name)["status"]["state"] == desired_state
 
+    def get_volume_state(self, volume_name):
+        logging.info(f"getting volume {volume_name} state")
+        return self.get(volume_name)["status"]["robustness"]
+
     def get_endpoint(self, volume_name):
-        warnings.warn("no endpoint in volume cr, get it from rest api")
+        logging.info(f"getting volume {volume_name} endpoint")
         return Rest(self.node_exec).get_endpoint(volume_name)
 
     def write_random_data(self, volume_name, size):
+        logging.info(
+            f"writing {size} Mb data into volume {volume_name} mount point")
+
         node_name = self.get(volume_name)["spec"]["nodeID"]
         endpoint = self.get_endpoint(volume_name)
+
+        logging.debug(
+            f"starting write data into {endpoint} on node {node_name}")
         checksum = self.node_exec.issue_cmd(
             node_name,
-            f"dd if=/dev/urandom of={endpoint} bs=1M count={size} status=none;\
+            f"dd if=/dev/urandom of={endpoint} bs=2M count={size} status=none;\
               md5sum {endpoint} | awk \'{{print $1}}\'")
+        logging.debug(f'finishing write data')
         return checksum
 
-    def delete_replica(self, volume_name, node_name):
-        replica_list = self.obj_api.list_namespaced_custom_object(
+    def get_replica(self, volume_name, node_name):
```

Comment:
Suggest segregating each logic related to the CR into its dedicated directory, such as "engine," "replica," etc.

Reply:
Got it.  For different CRs, move them to their respective class file definitions.
```
├── libs
│   ├── engine
│   │   ├── __init__.py
│   │   ├── base.py
│   │   ├── crd.py
│   │   └── engine.py
│   ├── node
│   │   ├── __init__.py
│   │   ├── aws.py
│   │   ├── local_cluster.py
│   │   └── node.py
│   ├── node_exec
│   │   ├── __init__.py
│   │   └── node_exec.py
│   ├── replica
│   │   ├── __init__.py
│   │   ├── base.py
│   │   ├── crd.py
│   │   ├── replica.py
│   │   └── rest.py
│   └── volume
│       ├── __init__.py
│       ├── base.py
│       ├── crd.py
│       ├── rest.py
│       └── volume.py
```